### PR TITLE
Add purge script

### PR DESCRIPTION
**Pull request by apentori**
Linked to : 

* https://github.com/status-im/infra-ci/issues/121

Depends on:

* https://github.com/status-im/infra-role-gitea/pull/3

Can be improve to set special depending on the name of the container if need be. In that case, it will need to read a custom yaml conf file.

### CODE REVIEW

File: defaults/main.yml [hunk 1]
```diff
@@ -56,6 +56,13 @@ gitea_organizations: []
 #  - org1 
 #  - org2
 
+# Purge
+gitea_purge_timer_name: 'gitea-purge-artifact'
+gitea_purge_timer_user: '{{ gitea_app_host_uid }}'
+gitea_purge_timer_frequency: 'daily'
+gitea_purge_timer_timeout: 120
+gitea_purge_max_retention: 15
```
Comment (jakubgs) [inline defaults/main.yml]: If this purge is only for containers then the variables should include `container` in the name:
```suggestion
# Purge
gitea_container_purge_timer_name: 'gitea-purge-artifact'
gitea_container_purge_timer_user: '{{ gitea_app_host_uid }}'
gitea_container_purge_timer_frequency: 'daily'
gitea_container_purge_timer_timeout: 120
gitea_container_purge_max_retention: 15
```

File: templates/purge-registry.py.j2 [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    if args.username is not None:
+        logging.info('Loading username from parameter')
+    if args.password is not None:
+        logging.warn('Loading password from parameter, not recommended')
+    if environ.get('GITEA_USERNAME') is not None:
+        logging.info('Loading username from environment variable')
+        username=environ.get('GITEA_USERNAME')
+    if environ.get('GITEA_USERNAME') is not None:
+        logging.info('Loading password from environment variable')
+        username=environ.get('GITEA_PASSWORD')
```
Comment (jakubgs) [inline templates/purge-registry.py.j2]: Same as in the other init script. Just use `default` and `assert` in `parse_args()`.

File: templates/purge-registry.py.j2 [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            return True
+        else:
+            logging.info('The artifact %s is %s old' % (self.name, abs((datetime.datetime.now() - self.creation_date))))
+            return False
+
+def get_call(url, basic):
+    logging.debug('HTTP GET call to %s' % url)
+    res=requests.get(url, auth=basic)
+    res.raise_for_status()
+    return res
```
Comment (jakubgs) [inline templates/purge-registry.py.j2]: You are repeating yourself too much. This needs to be a `GiteaAPI` class that has thin-wrappers around `_request()`.

Like this:
https://github.com/status-im/infra-utils/blob/34e5d068a46c408426b7b6c4182f1ab828632528/ethereum/validator_check.py#L12-L34

File: templates/purge-registry.py.j2 [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    for org in orgs:
+        logging.info('Get containers from %s' % org)
+        try:
+            conts = get_artifacts(args.url, basic, org)
+            for cont in conts:
+                if cont.should_be_purge(retention):
+                    try:
+                        delete_artifact(url, basic, cont)
+                     except requests.exceptions.HTTPError as err:
+                        failed.append(err.response.text)
```
Comment (jakubgs) [inline templates/purge-registry.py.j2]: Try avoiding too many nesting indents. You can do that by inverting `if`s:
```suggestion
                if not cont.should_be_purge(retention):
                    continue
                try:
                    delete_artifact(url, basic, cont)
                 except requests.exceptions.HTTPError as err:
                    failed.append(err.response.text)
```

File: templates/purge-registry.py.j2 [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+def main():
+    args = parse_args()
+    logging.basicConfig(level=args.log.upper())
+    basic = prepare_credentials(args)
+    retention=int(args.retention)
+    orgs=[]
+    try:
+        orgs=get_organization(args.url, basic)
+    except request.exceptions.HTTPError as err:
+        exit(1)
```
Comment (jakubgs) [inline templates/purge-registry.py.j2]: What is the point of this `try`/`except`? Why are you hiding an exception? Just let it bubble to the top and break.

File: templates/purge-registry.py.j2 [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    parser.add_argument('-p', '--password',   required=False,  help='Password')
+    parser.add_argument('-l', '--log',        required=False, default='INFO')
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+    logging.basicConfig(level=args.log.upper())
+    basic = prepare_credentials(args)
+    retention=int(args.retention)
```
Comment (jakubgs) [inline templates/purge-registry.py.j2]: Just apply `type=int` to the `add_argument()` method.

**Review by jakubgs — CHANGES_REQUESTED** [review]

**Comment by jakubgs** [comment]
Some issues I saw:
```
  File "/docker/gitea/purge-registry.py", line 92
    parser.add_argument('-r', '--retention',  required=True,  type=int, help='Maximum retention period for each artifact in days', type=int)
                                                                                                                                   ^^^^^^^^
SyntaxError: keyword argument repeated: type
```
```
Traceback (most recent call last):
  File "/docker/gitea/purge-registry.py", line 124, in <module>
    main()
  File "/docker/gitea/purge-registry.py", line 104, in main
    giteaApi = GiteaApi('%s/api/v1' % args.url, args.username, args.password)
  File "/docker/gitea/purge-registry.py", line 42, in __init__
    self.auth = requests.auth.HTTPBasicAuth(args.username, args.password)
NameError: name 'args' is not defined
```
```
Traceback (most recent call last):
  File "/docker/gitea/purge-registry.py", line 124, in <module>
    main()
  File "/docker/gitea/purge-registry.py", line 105, in main
    orgs=get_organization(giteaApi)
  File "/docker/gitea/purge-registry.py", line 62, in get_organization
    res=giteaApi._get('orgs')
  File "/docker/gitea/purge-registry.py", line 53, in _get
    return self._req('GET', path)
  File "/docker/gitea/purge-registry.py", line 49, in _req
    logging.error(tval.text)
NameError: name 'tval' is not defined. Did you mean: 'rval'?
```
```
ERROR:root:Error 404 when trying to get the organization: Not found.
Traceback (most recent call last):
  File "/docker/gitea/purge-registry.py", line 124, in <module>
    main()
  File "/docker/gitea/purge-registry.py", line 107, in main
    for org in orgs:
TypeError: 'NoneType' object is not iterable
```

**Comment by apentori** [comment]
Issues solved:

```
Nov 27 12:53:37 node-01.do-ams3.harbor.ci systemd[1]: Starting Timer to perform clean...
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:Querrying all the organization
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:Deployment : ['acid-info', 'nimbus', 'status', 'wakuorg']
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:Get containers from acid-info
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:Querrying org acid-info to list all artifacts
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/relayer is 5 days, 4:01:33.884066 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/relayer is 5 days, 4:01:42.884304 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/contracts is 5 days, 4:01:48.884376 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/relayer is 5 days, 4:01:54.884420 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/ui is 5 days, 4:01:58.884458 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/contracts is 5 days, 4:01:59.884495 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/contracts is 5 days, 4:02:18.884530 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/ui is 5 days, 4:02:18.884563 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/relayer is 5 days, 4:02:22.884596 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/contracts is 5 days, 4:02:36.884631 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/ui is 5 days, 4:02:41.884665 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/relayer is 5 days, 4:02:56.884698 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/contracts is 5 days, 4:02:58.884732 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/ui is 5 days, 4:03:05.884765 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/contracts is 5 days, 4:03:17.884803 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/relayer is 5 days, 4:03:27.884835 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/ui is 5 days, 4:03:34.884869 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/contracts is 5 days, 4:03:38.884903 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/relayer is 5 days, 4:04:00.884938 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/ui is 5 days, 4:04:01.884973 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/contracts is 5 days, 4:04:01.885006 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/relayer is 5 days, 4:04:24.885042 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/ui is 5 days, 4:04:24.885076 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/contracts is 5 days, 4:04:24.885103 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/ui is 5 days, 4:04:48.885128 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/relayer is 5 days, 4:04:48.885152 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact kurate/contracts is 5 days, 4:04:49.885176 old
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:Get containers from nimbus
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:Querrying org nimbus to list all artifacts
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:Get containers from status
Nov 27 12:53:37 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:Querrying org status to list all artifacts
Nov 27 12:53:38 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:Get containers from wakuorg
Nov 27 12:53:38 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:Querrying org wakuorg to list all artifacts
Nov 27 12:53:38 node-01.do-ams3.harbor.ci purge-registry.py[2871158]: INFO:root:The artifact redis is 5 days, 4:02:22.104223 old
Nov 27 12:53:38 node-01.do-ams3.harbor.ci systemd[1]: gitea-purge-artifact.service: Deactivated successfully.
Nov 27 12:53:38 node-01.do-ams3.harbor.ci systemd[1]: Finished Timer to perform clean.
```

File: tasks/purge-registry.yml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    systemd_timer_name: '{{ gitea_container_purge_timer_name }}'
+    systemd_timer_description: 'Timer to perform clean '
+    systemd_timer_dynamic_user: true
+    systemd_timer_frequency: '{{ gitea_container_purge_timer_frequency }}'
+    systemd_timer_timeout_sec: '{{ gitea_container_purge_timer_timeout }}'
+    systemd_timer_work_dir: '{{ gitea_service_path }}'
+    systemd_timer_requires_extra: 'docker.service'
+    systemd_timer_start_on_creation: false
+    systemd_timer_script_path: '{{ gitea_service_path}}/purge-registry.py'
+    systemd_timer_script_args: '--url http://localhost:{{ gitea_app_cont_web_port }}  --retention {{ gitea_purge_max_retention }}'
```
Comment (jakubgs) [inline tasks/purge-registry.yml]: You can make this more readable using `>-`:
```suggestion
    systemd_timer_script_args: >-
      --url http://localhost:{{ gitea_app_cont_web_port }}
      --retention {{ gitea_purge_max_retention }}
```

**Review by jakubgs — APPROVED** [review]
Looks good.

**Pull request merged**